packages/cw0/src/macros.rs

```diff
@@ -0,0 +1,94 @@
+/// Quick check for a guard. If the condition (first argument) is false,
+/// then return the second argument wrapped in Err(x).
+///
+///   ensure!(permissions.delegate, ContractError::DelegatePerm {});
+///  is the same as
+///   if !permissions.delegate {
+///     return Err(ContractError::DelegatePerm {});
+///   }
+#[macro_export]
+macro_rules! ensure {
+    ($cond:expr, $e:expr) => {
+        if !($cond) {
+            return Err($e);
+        }
+    };
+}
+
+/// Opposite of ensure. If the condition (first argument) is true,
+/// then return the second argument wrapped in Err(x).
+///
+///   fail_if!(!permissions.delegate, ContractError::DelegatePerm {});
+///  is the same as
+///   if !permissions.delegate {
+///     return Err(ContractError::DelegatePerm {});
+///   }
+#[macro_export]
+macro_rules! fail_if {
+    ($cond:expr, $e:expr) => {
+        if ($cond) {
+            return Err($e);
+        }
+    };
+}
+
+/// Quick check for a guard. Like assert_eq!, but rather than panic,
+/// it returns the second argument wrapped in Err(x).
+///
+///   ensure_eq!(info.sender, cfg.admin, ContractError::Unauthorized {});
+///  is the same as
+///   if info.sender != cfg.admin {
+///     return Err(ContractError::Unauthorized {});
+///   }
+#[macro_export]
+macro_rules! ensure_eq {
+    ($a:expr, $b:expr, $e:expr) => {
+        if $a != $b {
+            return Err($e);
```

Comment:
Now what if we add

```suggestion
            return Err($e.into());
```

Wouldn't that open up a huge space of use cases in a very natural way?

Reply:
I thought this... but then it requires you to explicitly define the output. Like in the closures in the tests. 

Maybe not being able to infer the type is worth it for the flexibility. 🤷 

Can you make a PR on top. I am especially interested in the ergonomics of the change in real contract usages, not just my silly test code. It may be able to infer all types in the contracts, so my hesitation is unwarranted.

Reply:
Will do. This would make this macro behave like `try!` and `?` when it comes to conversion where you have the same restriction on inference. See

```rust
#[macro_export]
#[stable(feature = "rust1", since = "1.0.0")]
#[rustc_deprecated(since = "1.39.0", reason = "use the `?` operator instead")]
#[doc(alias = "?")]
macro_rules! r#try {
    ($expr:expr $(,)?) => {
        match $expr {
            $crate::result::Result::Ok(val) => val,
            $crate::result::Result::Err(err) => {
                return $crate::result::Result::Err($crate::convert::From::from(err));
            }
        }
    };
}
```